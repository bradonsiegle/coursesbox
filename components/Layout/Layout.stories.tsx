import { ComponentStoryObj, ComponentMeta } from "@storybook/react";
import { screen, userEvent } from "@storybook/testing-library";

import { Layout } from "./Layout";

export default {
  title: "Layout/Layout",
  component: Layout,
} as ComponentMeta<typeof Layout>;

export const BasicLayout: ComponentStoryObj<typeof Layout> = {
  play: async ({ args }) => {
    await userEvent.click(screen.getByRole("button"));
  },
  args: {
    children: (
      <>
        <h1>Layout</h1>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed euismod,
          nisl nec ultricies lacinia, nisl nisl aliquet nisl, vitae aliquam nisl
          nisl sit amet lorem. Nulla facilisi. Sed euismod, nisl nec ultricies
          lacinia, nisl nisl aliquet nisl, vitae aliquam nisl nisl sit amet
          lorem. Nulla facilisi. Sed euismod, nisl nec ultricies lacinia, nisl
          nisl aliquet nisl, vitae aliquam nisl nisl sit amet lorem. Nulla
          facilisi. Sed euismod, nisl nec ultricies lacinia, nisl nisl aliquet
          nisl, vitae aliquam nisl nisl sit amet lorem. Nulla facilisi. Sed
        </p>
      </>
    ),
  },
};
