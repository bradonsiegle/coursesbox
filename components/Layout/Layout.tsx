import { FC } from 'react';
import styled from '@emotion/styled';

import { Logo } from '@/components/Logo';
