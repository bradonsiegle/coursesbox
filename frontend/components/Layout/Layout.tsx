import Link from "next/link";
import { useSelector, useDispatch } from "react-redux";
import { AppDispatch, RootState } from "@/store";
import { FC, useState, useEffect, useLayoutEffect } from "react";
import { IconButton } from "@/components/IconButton";
import { StyledLink } from "../StyledLink";
import { ThemeProvider } from "@emotion/react";
import { Themes } from "@/styles/themes";
import { login, selectUser } from "@/services/userSlice";

import {
  Wrapper,
  LogoLink,
  StyledLogo,
  MainNav,
  SearchInput,
  Content,
  Footer,
} from "./components";

interface Props {
  children?: React.ReactNode;
}

const useIsomorphicLayoutEffect =
  typeof window !== "undefined" ? useLayoutEffect : useEffect;

export const Layout: FC<Props> = ({ children }) => {
  const { username } = useSelector<RootState, RootState["user"]>(selectUser);

  const [isDark, setIsDark] = useState(true);
  const dispatch = useDispatch<AppDispatch>();

  const toggleDark = () => {
    localStorage.setItem("theme", isDark ? "light" : "dark");
    setIsDark(!isDark);
  };

  useIsomorphicLayoutEffect(() => {
    dispatch(login());

    const theme = localStorage.getItem("theme");
    const themeExistsInStorage = Boolean(theme !== null);

    const isDark = themeExistsInStorage
      ? Boolean(theme === "dark")
      : window.matchMedia("(prefers-color-scheme: dark)").matches;
    setIsDark(isDark);
  }, []);

  const theme = Themes[isDark ? "dark" : "light"];

  return (
    <ThemeProvider theme={theme}>
      <Wrapper>
        <Link href="/" passHref>
          <LogoLink>
            <StyledLogo size={3}>
              <span className="logo_short">CBOX</span>
              <span className="logo_long">CoursesBox</span>
            </StyledLogo>
          </LogoLink>
        </Link>
        <MainNav>
          <Link href="/all" passHref>
            <StyledLink>All</StyledLink>
          </Link>

          <Link href={username ? "/user" : "/login"} passHref>
            <IconButton name={username ? "User" : "Login"} size={1} />
          </Link>

          <IconButton
            name={!isDark ? "Moon" : "Sun"}
            size={1}
            onClick={toggleDark}
          />
        </MainNav>
        <SearchInput icon="Search" placeholder="Search" onChange={() => null} />
        <Content>{children}</Content>
        <Footer>
          <p>Â© 2022 Bradon Siegle. All rights reserved.</p>
        </Footer>
      </Wrapper>
    </ThemeProvider>
  );
};
