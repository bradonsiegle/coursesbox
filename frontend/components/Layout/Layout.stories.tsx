import { ComponentStoryObj, ComponentMeta } from "@storybook/react";
import { screen, userEvent } from "@storybook/testing-library";
import { Provider } from "react-redux";
import { store } from "@/store";

import { Layout } from "./Layout";

export default {
  title: "Layout/Layout",
  component: Layout,
} as ComponentMeta<typeof Layout>;

export const BasicLayout: ComponentStoryObj<typeof Layout> = {
  args: {
    children: (
      <>
        <h1>Layout</h1>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed euismod,
          nisl nec ultricies lacinia, nisl nisl aliquet nisl, vitae aliquam nisl
          nisl sit amet lorem. Nulla facilisi. Sed euismod, nisl nec ultricies
          lacinia, nisl nisl aliquet nisl, vitae aliquam nisl nisl sit amet
          lorem. Nulla facilisi. Sed euismod, nisl nec ultricies lacinia, nisl
          nisl aliquet nisl, vitae aliquam nisl nisl sit amet lorem. Nulla
          facilisi. Sed euismod, nisl nec ultricies lacinia, nisl nisl aliquet
          nisl, vitae aliquam nisl nisl sit amet lorem. Nulla facilisi. Sed
        </p>
      </>
    ),
  },
  decorators: [
    (Story) => (
      <Provider store={store}>
        <Story />
      </Provider>
    ),
  ],
};
