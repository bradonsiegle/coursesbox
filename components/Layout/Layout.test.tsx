import { render } from '@/test-utils';

import { Layout } from './Layout';

describe('Layout test cases', () => {
	const child = (
		<>
			<h1>Main Article</h1>
			<p>Some text</p>
		</>
	);

	it('should render the layout', () => {
		const { container } = render(
			<Layout isDark onThemeToggle={() => undefined}>
				{child}
			</Layout>
		);
		expect(container).toMatchSnapshot();
	});
});
