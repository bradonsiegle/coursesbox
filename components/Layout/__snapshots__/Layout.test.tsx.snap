// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout test cases should render the layout 1`] = `
<div>
  <div
    class="css-1d94mxe"
  >
    <a
      class="css-1695qjj"
      href="/"
    >
      <header
        class="css-1fzjdke"
        size="3"
      >
        C8X
      </header>
    </a>
    <nav
      class="css-tkz1x8"
    >
      <a
        href="/all"
      >
        All
      </a>
      <a
        href="/news"
      >
        News
      </a>
      <button
        class="css-25hrr4"
        title="Sun"
      >
        <svg
          aria-label="Sun"
          class="css-hxs6wy"
          fill="none"
          height="1rem"
          role="img"
          viewBox="0 0 24 24"
          width="1rem"
        >
          <path
            clip-rule="evenodd"
            d="M12 16a4 4 0 100-8 4 4 0 000 8zm0 2a6 6 0 100-12 6 6 0 000 12zM11 0h2v4.062a8.079 8.079 0 00-2 0V0zM7.094 5.68L4.222 2.808 2.808 4.222 5.68 7.094A8.048 8.048 0 017.094 5.68zM4.062 11H0v2h4.062a8.079 8.079 0 010-2zm1.618 5.906l-2.872 2.872 1.414 1.414 2.872-2.872a8.048 8.048 0 01-1.414-1.414zM11 19.938V24h2v-4.062a8.069 8.069 0 01-2 0zm5.906-1.618l2.872 2.872 1.414-1.414-2.872-2.872a8.048 8.048 0 01-1.414 1.414zM19.938 13H24v-2h-4.062a8.069 8.069 0 010 2zM18.32 7.094l2.872-2.872-1.414-1.414-2.872 2.872c.528.41 1.003.886 1.414 1.414z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </nav>
    <label
      class="css-1ikem8n"
      height="7"
      width="20"
    >
      <div
        class="css-1igl99e"
      >
        <input
          class="css-13slppn"
          placeholder="Search"
        />
        <svg
          aria-hidden="true"
          aria-label="Search"
          class="css-7drimr"
          data-icon="search"
          fill="currentColor"
          focusable="false"
          height="2rem"
          role="img"
          viewBox="64 64 896 896"
          width="2rem"
        >
          <path
            d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
          />
        </svg>
      </div>
    </label>
    <main
      class="css-6z89iq"
    >
      <h1>
        Main Article
      </h1>
      <p>
        Some text
      </p>
    </main>
    <footer
      class="css-1wymfr2"
    >
      <p>
        Â© 2022 Bradon Siegle. All rights reserved.
      </p>
    </footer>
  </div>
</div>
`;
